src/main/java/io/neonbee/internal/WriteSafeRegistry.java

```diff
@@ -0,0 +1,149 @@
+
+    /**
+     * Register the value in the {@link NeonBee} async shared map under the sharedMapKey.
+     *
+     * @param sharedMapKey      the shared map key
+     * @param value             the value to register
+     * @param valueAddedHandler a handler to handle additional things when the value is registered in the map. The
+     *                          handler isn't executed if the value is already registered in the shared map.
+     * @return the future
+     */
+    public Future<Void> register(String sharedMapKey, Object value, Handler<Promise<Void>> valueAddedHandler) {
+        logger.info("register value: \"{}\" in shared map: \"{}\"", sharedMapKey, value);
+
+        return lock(sharedMapKey, () -> addValue(sharedMapKey, value, valueAddedHandler));
+    }
+
+    /**
+     * Method that acquires a lock for the sharedMapKey and released the lock after the futureSupplier is executed.
+     *
+     * @param sharedMapKey   the shared map key
+     * @param futureSupplier supplier for the future to be secured by the lock
+     * @return the futureSupplier
+     */
+    private Future<Void> lock(String sharedMapKey, Supplier<Future<Void>> futureSupplier) {
+        logger.debug("Get lock for {}", sharedMapKey);
+        return sharedDataAccessor.getLock(sharedMapKey).onFailure(throwable -> {
+            logger.error("Error acquiring lock for {}", sharedMapKey, throwable);
+        }).compose(lock -> futureSupplier.get().onComplete(anyResult -> {
+            logger.debug("Releasing lock for {}", sharedMapKey);
+            lock.release();
+        }));
+    }
+
+    private Future<Void> addValue(String sharedMapKey, Object value, Handler<Promise<Void>> valueAddedHandler) {
+        Future<AsyncMap<String, Object>> sharedMap = getSharedMap();
+
+        return sharedMap.compose(map -> map.get(sharedMapKey))
+                .map(valueOrNull -> valueOrNull != null ? (JsonArray) valueOrNull : new JsonArray())
+                .compose(valueArray -> {
+                    Future<Void> future = succeededFuture();
+                    if (!valueArray.contains(value)) {
+                        valueArray.add(value);
+                        future = Future.future(valueAddedHandler);
+                    }
+
+                    if (logger.isInfoEnabled()) {
+                        logger.info("Registered verticle {} of node {} in shared map.", value,
+                                NeonBee.get(vertx).getNodeId());
+                    }
+
+                    return CompositeFuture.all(sharedMap.compose(map -> map.put(sharedMapKey, valueArray)), future)
+                            .mapEmpty();
+                });
+    }
+
+    /**
+     * Unregister the value in the {@link NeonBee} async shared map from the sharedMapKey.
+     *
+     * @param sharedMapKey the shared map key
+     * @param value        the value to register
+     * @return the future
+     */
+    public Future<Void> unregister(String sharedMapKey, String value) {
```

Comment:
this method is completely untested. Also noticed there is no `WriteSafeRegistryTest` class; might be a good idea to introduce and write a test for `unregister`.